### PR TITLE
OpenStack Keystone V3 - different auth "domains" support (JCLOUDS-1414)

### PR DESCRIPTION
**Pull request by axel3rd**
Cf. [JCLOUDS-1414](https://issues.apache.org/jira/browse/JCLOUDS-1414) ; linked with [jclouds-site#214](https://github.com/jclouds/jclouds-site/pull/214)

### CODE REVIEW

File: apis/openstack-keystone/src/main/java/org/jclouds/openstack/keystone/v3/binders/BindAuthToJsonPayload.java [hunk 1]
```diff
@@ -93,15 +93,29 @@ private Object parseScope(TenantOrDomainAndCredentials<T> credentials) {
...
+         // backward compatibility
+         return ProjectIdScope.create(ProjectId.create(parts[1], parseProjectDomain(credentials, false)));
       } else if (DOMAIN.equals(parts[0])) {
          return DomainScope.create(Name.create(parts[1]));
       } else {
          return DomainIdScope.create(Id.create(parts[1]));
       }
    }
+
+   private Object parseProjectDomain(TenantOrDomainAndCredentials<T> credentials, boolean useTenant) {
```
Comment (nacx) [inline apis/openstack-keystone/src/main/java/org/jclouds/openstack/keystone/v3/binders/BindAuthToJsonPayload.java]: Do we really need this boolean `useTenant` flag? Could we just change precedence and check the new properties first, then fallback to the tenant ones?

File: apis/openstack-keystone/src/test/java/org/jclouds/openstack/keystone/v3/auth/V3AuthenticationApiMockTest.java [hunk 2]
```diff
@@ -31,71 +32,118 @@
...
-      AuthInfo authInfo = authenticationApi.authenticatePassword(credentials);
 
-      assertTrue(authInfo instanceof Token);
-      assertEquals(authInfo, tokenFromResource("/v3/token.json"));
+      checkTokenResult(credentials, "/v3/auth-password-scoped.json");
+   }
 
-      assertEquals(server.getRequestCount(), 1);
-      assertSent(server, "POST", "/auth/tokens", stringFromResource("/v3/auth-password-scoped.json"));
+   public void testAuthenticatePasswordProjectScopedNameDomainHack() throws InterruptedException {
```
Comment (nacx) [inline apis/openstack-keystone/src/test/java/org/jclouds/openstack/keystone/v3/auth/V3AuthenticationApiMockTest.java]: `s/Hack/BackwardsCompat/` ?

File: apis/openstack-keystone/src/test/java/org/jclouds/openstack/keystone/v3/auth/V3AuthenticationApiMockTest.java [hunk 3]
```diff
@@ -31,71 +32,118 @@
...
+      if (credentials.credentials() instanceof PasswordCredentials) {
+         authInfo = authenticationApi
+               .authenticatePassword((TenantOrDomainAndCredentials<PasswordCredentials>) credentials);
+      } else if (credentials.credentials() instanceof TokenCredentials) {
+         authInfo = authenticationApi.authenticateToken((TenantOrDomainAndCredentials<TokenCredentials>) credentials);
+      } else if (credentials.credentials() instanceof ApiAccessKeyCredentials) {
+         authInfo = authenticationApi
+               .authenticateAccessKey((TenantOrDomainAndCredentials<ApiAccessKeyCredentials>) credentials);
+      } else {
+         throw new InterruptedException(String.format("Unsupported authentication method with class: %s",
```
Comment (nacx) [inline apis/openstack-keystone/src/test/java/org/jclouds/openstack/keystone/v3/auth/V3AuthenticationApiMockTest.java]: Better throw an `IllegalArgumentException`.

File: apis/openstack-keystone/src/main/java/org/jclouds/openstack/keystone/v3/binders/BindAuthToJsonPayload.java (hunk 1 above)
Comment (axel3rd) [reply apis/openstack-keystone/src/main/java/org/jclouds/openstack/keystone/v3/binders/BindAuthToJsonPayload.java]: For project/id, there is [currently](https://github.com/jclouds/jclouds/blob/master/apis/openstack-keystone/src/main/java/org/jclouds/openstack/keystone/v3/binders/BindAuthToJsonPayload.java#L100) no domain created on tenant properties => it was more for backwardcompatibility (use new properties only for project/id, both for project/name).

**Comment by axel3rd** [comment]
@nacx : I will do that (your reviews, thanks) this wednesday. About `useTenant` flag I will be more explicit and add a unit test to prove the use case.

**Comment by axel3rd** [comment]
> I will do that (your reviews, thanks) this wednesday. About useTenant flag I will be more explicit and add a unit test to prove the use case.

Done, with `useTenant` renamed with [some explanation](https://github.com/jclouds/jclouds/pull/1204/files#diff-918e1e161307a179875339dfa9caceedR109), with [dedicated unit test](https://github.com/jclouds/jclouds/pull/1204/files#diff-59311ec5e3d12e89d93b3e4bf1dc805dR52) for backwards compatibility.
Available for any review complement(s).

**Comment by nacx** [comment]
Pushed to [master](http://git-wip-us.apache.org/repos/asf/jclouds/commit/053dfd01) and [2.1.x](http://git-wip-us.apache.org/repos/asf/jclouds/commit/93a805ca). Thanks @axel3rd!

**Pull request closed without merging**